src/shuffler-codereview.c

```diff
@@ -0,0 +1,1228 @@
+        }
+        // shuffle the indicies
+        for (int i = totalPoolSize - 1; i > 0; i--) {
+            int r = tinymt32_generate_uint32(&masterSeed) % (i + 1);
+            int tmp;
+            SWAP(filteredIndicies[i], filteredIndicies[r], tmp);
+        }
+        // take the top 6 values from the newly shuffled list of indicies.
+        for (int i = 0; i < 6; i++) {
+            partyIndicies[i] = filteredIndicies[i];
+        }
+        break;
+    }
+    case TB_STRAT_GYM_MEMBER: {
+        trainer->partySize = 4;
+        // figure out which pool is largest so we can allocate our array to fit all the indicies.
+        int maxPoolSize = tt->pool1Size;
+        if (maxPoolSize < tt->pool2Size) {
+            maxPoolSize = tt->pool2Size;
+        }
+        if (maxPoolSize < tt->pool3Size) {
+            // I don't think this should ever happen; pool3 should necessarily
+            // be smaller than pools 1 and 2. But just in case...
+            maxPoolSize = tt->pool3Size;
+        }
+
+        int filteredPoolSize = tt->pool1Size;
+        int filteredIndicies[maxPoolSize];
+        int nextIndex = 0;
+        // we start with pool1. we want to pick 3 pokemon total from pool1.
+        // filter out possible pokemon that are of the wrong level.
+        for (int i = 0; i < tt->pool1Size; i++) {
+            if (!PokemonIsValidLevel(&tt->pool1[i], level)) {
+                filteredPoolSize--;
+                continue;
+            }
+            filteredIndicies[nextIndex] = i;
+            nextIndex++;
+        }
+        // shuffle the remaining indicies.
+        for (int i = filteredPoolSize - 1; i > 0; i--) {
+            int r = tinymt32_generate_uint32(&masterSeed) % (i + 1);
+            int tmp;
+            SWAP(filteredIndicies[i], filteredIndicies[r], tmp);
+        }
+        // fill the first 3 party slots with pokemon from pool1.
+        for (int i = 0; i < trainer->partySize - 1; i++) {
+            partyIndicies[i] = filteredIndicies[i];
+        }
+
+        // now we do the same for pool2, except we're only picking 1.
+        filteredPoolSize = tt->pool2Size;
+        nextIndex = 0;
+        // filter out possible pokemon that are of the wrong level.
+        for (int i = 0; i < tt->pool2Size; i++) {
+            if (!PokemonIsValidLevel(&tt->pool2[i], level)) {
+                filteredPoolSize--;
+                continue;
+            }
+            filteredIndicies[nextIndex] = i;
+            nextIndex++;
+        }
+        // shuffle the remaining indicies.
+        for (int i = filteredPoolSize - 1; i > 0; i--) {
+            int r = tinymt32_generate_uint32(&masterSeed) % (i + 1);
+            int tmp;
+            SWAP(filteredIndicies[i], filteredIndicies[r], tmp);
+        }
+        // put a pokemon from pool2 into the final party slot.
+        partyIndicies[trainer->partySize - 1] = filteredIndicies[0] + tt->pool1Size;
+
+        // if there is no pool3, we're done.
```

Comment:
I feel like I'm reading the same piece of code three times. I can't prove it, but it feels that way, and it feels like it should be a function.